src/python/pants/version.py

```diff
@@ -56,11 +56,32 @@ def __ge__(self, other: Any):
 _PANTS_VERSION_OVERRIDE = "_PANTS_VERSION_OVERRIDE"
 
 
+# @TODO: Use https://github.com/pypa/setuptools_scm/pull/1005 when we can
+def _determine_version_from_pants_source():
+    from setuptools_scm import get_version  # pants: no-infer-dep
+
+    def version_scheme(version):
+        return version.format_with("{tag}")
+
+    val = get_version(
+        version_scheme=version_scheme,
+        tag_regex=r"^release_(?P<version>[vV]?\d+(?:\.\d+){0,2}[^\+]*)$",
+    )
+    return val
+
+
 VERSION: str = (
     # Do not remove/change this env var without coordinating with `pantsbuild/scie-pants` as it is
     # being used when bootstrapping Pants with a released version.
     os.environ.get(_PANTS_VERSION_OVERRIDE)
     or
+    # NB: This is only relevant for the Pants repo itself
+    (
+        _determine_version_from_pants_source()
```

Comment:
Oh and I guess to answer your questions. I'd rather err on the side of caution. Where we could reasonably ask users to run `--version` and it gives us something useful.

Then, if users want to exist outside of this model, they can do so by either setting the env var, or modifying the file.

Reply:
I'm also toying with speeding up option registration by doing it at build time instead of runtime, and in that case it would have to be tied to a specific version, and it would be chaotic to not invalidate that correctly.

Reply:
Just so I'm clear, is that something that'll preclude and block this PR? Or is it something that'll be handled in yours? Or something else?


Just trying to understand the impact of your statement.

Reply:
I'm just providing a reason that every version should have a well-defined version, even when running from sources, rather than using a fixed dummy.